sdk/src/rhesis/sdk/services/context_generator.py

```
@@ -0,0 +1,192 @@
+"""Context generator service for creating context from various sources."""
+
+import re
+from typing import List
+
+from rhesis.sdk.utils import count_tokens
+
+
+class ContextGenerator:
+    """Service for generating context from various sources using intelligent semantic chunking."""
+
+    def __init__(
+        self,
+        max_context_tokens: int = 1500,  # User's preferred max
+        absolute_max_context_tokens: int = 2500,  # Hard safety limit
+    ):
+        """
+        Initialize the context generator.
+
+        Args:
+            max_context_tokens: Maximum tokens per context (user preference)
+            absolute_max_context_tokens: Hard limit that cannot be exceeded
+        """
+        self.max_context_tokens = max_context_tokens
+        self.absolute_max_context_tokens = absolute_max_context_tokens
+
+        # Validate user input
+        if self.max_context_tokens > self.absolute_max_context_tokens:
+            raise ValueError(
+                f"max_context_tokens ({max_context_tokens}) cannot exceed "
+                f"absolute_max_context_tokens ({absolute_max_context_tokens})"
+            )
+
+    def generate_contexts(self, text: str) -> List[str]:
+        """
+        Generate contexts using intelligent semantic chunking with hard size limits.
+
+        Strategy:
+        1. Identify semantic boundaries (headers, sections, paragraphs)
+        2. Create contexts that respect these boundaries
+        3. Enforce hard token limit; if a semantic span exceeds the context limit, split it abruptly
+        4. If there are no internal boundaries, slice the text into token-capped windows
+        """
+        if not text:
+            return []
```

If the stirng is empty it should raise an error